### ISSUE TITLE
feat: configurable OpenAI API endpoint.

### ISSUE THREAD

**Issue by ahkhan03** [ollama-webui/ollama-webui#230]
**Is your feature request related to a problem? Please describe.**
Make the API endpoint url configurable so the user can connect other OpenAI-compatible APIs with the web-ui.

**Describe the solution you'd like**
Make it configurable through environment variables or add a new field in the **Settings** > **Add-ons**.

**Describe alternatives you've considered**
Tried [this](https://www.typingmind.com/) but it is not open source and self-hostable.

**Additional context**
Nill.

**Comment by justinh-rahb** [comment]
Would love to see this as well. Supposedly the new Mistral API platform is OpenAI API compatible, so that too would be an option available that I'm personally most interested in right now. In addition to Azure OpenAPI endpoints, LiteLLM, etc.

**Comment by tjbck** [comment]
Actively taking a look! One question I have is, how would users fetch the list of available models? Do other services offer api endpoints equivalent to `https://api.openai.com/v1/models`?

**Comment by Flowm** [comment]
The Mistral API seems to be very similar ([Docs](https://docs.mistral.ai/), [API spec](https://docs.mistral.ai/api/))

**Comment by tjbck** [comment]
@Flowm, Could you try to check if Mistral api supports the `/models` api route? Thanks!

**Comment by ahkhan03** [comment]
I tried it with mistral and anyscale, and both support `/models` route. I tried several others in the past few days, and all of them supported it, but I have confirmed for these two just now. They have additional fields in the returned JSON, but they all have `data.id`, `data.object`, and `data.owned_by` in common, which should be enough for integrating with OpenAI compatible tools.

**Comment by justinh-rahb** [comment]
@tjbck I'd be happy to give you an API key (with an expiry date of course) for Mistral if it'll help you get this feature completed. You can find my contact info in my profile.

**Comment by chymian** [comment]
to name one more on the ever growing list of openAI compatible APIs with resonable pricing: __together.ai__

API-endpoint: https://api.together.xyz/

they offer a $25 entry bonus and prices for 1M token between $0.1 to $0.9.

**Comment by tjbck** [comment]
Hey guys! Just added the custom api endpoint feature to #265 branch. If any of you guys could try it out and confirm that it works, I'll merge it to main! Thanks!

**Comment by justinh-rahb** [comment]
> Hey guys! Just added the custom api endpoint feature to #265 branch. If any of you guys could try it out and confirm that it works, I'll merge it to main! Thanks!

@tjbck Loading the model names from the /models endpoint is working:
<img width="778" alt="Screenshot 2023-12-22 at 11 55 51 PM" src="https://github.com/ollama-webui/ollama-webui/assets/52832301/cee2d382-0e7c-4a37-a7da-30e2e7ad0705">

However, generation is not responding:
<img width="778" alt="Screenshot 2023-12-22 at 11 58 15 PM" src="https://github.com/ollama-webui/ollama-webui/assets/52832301/d9f4d46d-4daa-4e8d-989f-5eb3f9dc1906">

I got the same response with all models. Unfortunately the Docker log doesn't show me anything useful, anywhere else I can look or enable debugging?

**Comment by ahkhan03** [comment]
@tjbck I just tried this branch. The model list loads correctly, but it shows a CORS policy issue in the browser console.

```Access to fetch at 'https://api.mistral.ai/v1/chat/completions' from origin 'http://localhost:8080' has been blocked by CORS policy: Request header field x-title is not allowed by Access-Control-Allow-Headers in preflight response.```

```POST https://api.mistral.ai/v1/chat/completions net::ERR_FAILED```

```
TypeError: Failed to fetch
at window.fetch (start.c756bce8.js:1:1402)
at X (3.92fc93bc.js:4:1895)
at async 3.92fc93bc.js:1:6227
at async Promise.all (index 0)
at async z (3.92fc93bc.js:1:6142)
at async ce (3.92fc93bc.js:7:2070)```

**Comment by tjbck** [comment]
Just a header problem it seems like, I'll remove some fields and let you guys know!

**Comment by tjbck** [comment]
Please try again and let me know!

**Comment by ahkhan03** [comment]
I tested it for anyscale and mistral models and it is working perfectly now.

**Comment by justinh-rahb** [comment]
I can archive my chatbot-ui forks now 😂

**Comment by tjbck** [comment]
@ahkhan03 @justinh-rahb Thanks a lot! Could not have implemented this feature without you guys 😅

**Comment by justinh-rahb** [comment]
@tjbck hoping to be able to contribute more, you've got the best UI going for Ollama **AND** ChatGPT/Mistral APIs right now. There's a lot of people getting interested in running local inference for office workers and solutions like this are key enablers. Here's to more in 2024! 🎉🚀

**Comment by davidsyoung** [comment]
Thank you for implementing this, HUGELY appreciated! 

By far the best UI. 

I am using it with TabbyAPI (Exllamav2 API), and unfort they set max_tokens to a default of 150.

Would it be possible in some way to get configurable max_tokens in the UI? It could be called num_predict to be correct for the modelfile, and if sending to an external openai use num_tokens instead?


Tried to use modelfile, but can’t set the “FROM” to an external model. This may be a better way of doing it?

FROM model_name:external

Would really appreciate it. Can’t get around 150 tokens for now!

**Comment by tjbck** [comment]
Hey guys! #381 migrates OpenAI API requests to the backend for proper multi-user support, I'd greatly appreciate if any one of you guys can help with the testing OpenAI compatible APIs, Thanks as always!

**Comment by ahkhan03** [comment]
I tested it with Anyscale and Mistral APIs, and it is working correctly for all users!!

**Comment by tjbck** [comment]
@ahkhan03 Thanks a lot!

**Comment by justinh-rahb** [comment]
Working with LiteLLM proxy for OpenAI, Mistral, Claude, etc. 👌
<img width="756" alt="Screenshot 2024-01-06 at 2 51 28 AM" src="https://github.com/ollama-webui/ollama-webui/assets/52832301/1d027ac4-23d7-410e-a90e-64d497c512dc">

**Comment by helenadeus** [comment]
hello - how did you configure claude as a model? do you have it locally or is there a way to connect to the anthropic apis..?

**Comment by OpenWaygate** [comment]
My endpoint don't support /models api, is there a way to call my endpoint directly? It's openai-compatible.

**Comment by Averyy** [comment]
You can add Claude by setting up litellm https://litellm.vercel.app/docs/proxy/deploy and then in Open WebUI you add your litellm ip/key in connections and it'll pull all your litellm. 

Took a bit of tinkering but once you get it setup it works perfectly.

**Comment by justinh-rahb** [comment]
> You can add Claude by setting up litellm https://litellm.vercel.app/docs/proxy/deploy and then in Open WebUI you add your litellm ip/key in connections and it'll pull all your litellm.
> 
> Took a bit of tinkering but once you get it setup it works perfectly.

Or use our [Pipelines](https://github.com/open-webui/pipelines) server, this is now our preferred method for Azure/Anthropic/Cohere/Google API access and much more.

**Comment by Averyy** [comment]
> > You can add Claude by setting up litellm https://litellm.vercel.app/docs/proxy/deploy and then in Open WebUI you add your litellm ip/key in connections and it'll pull all your litellm.
> > Took a bit of tinkering but once you get it setup it works perfectly.
> 
> Or use our [Pipelines](https://github.com/open-webui/pipelines) server, this is now our preferred method for Azure/Anthropic/Cohere/Google API access and much more.

Oh I didn't even realize this was an alternative, thanks will check it out.

**Comment by stanus74** [comment]
> Thank you for implementing this, HUGELY appreciated!
> 
> By far the best UI.
> 
> I am using it with TabbyAPI (Exllamav2 API), and unfort they set max_tokens to a default of 150.
> 
> Would it be possible in some way to get configurable max_tokens in the UI? It could be called num_predict to be correct for the modelfile, and if sending to an external openai use num_tokens instead?
> 
> Edit: Implemented in #322 along with some small fixes. :)

i can use tabbyapi with curl. i want use with openwebui . what is the endpoint. http://ip:8080  and set auth code dont work for me
